faas/request.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2021 Nitric Technologies Pty Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package faas
+
+type DataRequest interface {
+	Data() []byte
+}
+
+type dataRequestImpl struct {
+	data []byte
+}
+
+func (d *dataRequestImpl) Data() []byte {
+	return d.data
+}
+
+type HttpRequest interface {
+	DataRequest
+	Method() string
+	Path() string
+	Query() map[string][]string
+	Headers() map[string][]string
+}
+
+type httpRequestImpl struct {
+	dataRequestImpl
+	method  string
+	path    string
+	query   map[string][]string
+	headers map[string][]string
+}
+
+func (h *httpRequestImpl) Method() string {
+	return h.method
+}
+
+func (h *httpRequestImpl) Path() string {
+	return h.path
+}
+
+func (h *httpRequestImpl) Query() map[string][]string {
+	return h.query
+}
+
+func (h *httpRequestImpl) Headers() map[string][]string {
+	return h.headers
+}
+
+type EventRequest interface {
+	DataRequest
+	Topic() string
+}
```

Comment:
Do we want to include the mimeType here, it is provided with the proto contract.

Reply:
Will include it for now. Would be good to test this (in most cases the imputation may calculate application/octet-stream, I suspect we might deprecate this field in future in favour of a more contract driven approach instead. E.g. using protocol buffers or similiar for serializing/deserializing event payloads instead. Which with our new shared code capability would would quite well (similiar to how the http FaaS worker works with trigger requests on its request bodies).